File: dotnet/DesignPatterns/StructuralDesignPatterns/FlyweightDesignPattern/FlyweightDesignPattern.Example1/Tree.cs
namespace FlyweightDesignPattern.Example1;

/// <summary>
/// Context class that contains extrinsic (unique) state.
/// Each Tree has its own position (x, y) but shares the TreeType flyweight.
/// This separation allows thousands of trees to share a few TreeType objects.
/// </summary>
public class Tree
{
    public int X { get; }
    public int Y { get; }
    private readonly ITreeType _type;

    public Tree(int x, int y, ITreeType type)
    {
        X = x;
        Y = y;
        _type = type;
    }

    public void Draw()
    {
        _type.Draw(X, Y);
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/FlyweightDesignPattern/FlyweightDesignPattern.Example1/src/Tree.ts
import { ITreeType } from './ITreeType';

// Context — holds extrinsic (unique) state + reference to flyweight
export class Tree {
  constructor(
    public readonly x: number,
    public readonly y: number,
    private readonly type: ITreeType,
  ) {}

  draw(): void {
    this.type.draw(this.x, this.y);
  }
}
